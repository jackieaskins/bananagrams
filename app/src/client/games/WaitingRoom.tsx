import {
  Button,
  HStack,
  Heading,
  Stack,
  Text,
  useToast,
} from "@chakra-ui/react";
import { useCallback, useMemo } from "react";
import { FaRegCopy } from "react-icons/fa6";
import { useGame } from "./GameContext";
import OpponentBoardPreview from "@/client/boards/OpponentBoardPreview";
import PlayerTable from "@/client/players/PlayerTable";
import { socket } from "@/client/socket";
import { PlayerStatus } from "@/types/player";
import { ClientToServerEventName } from "@/types/socket";

export default function WaitingRoom(): JSX.Element {
  const {
    gameInfo: { gameName, players, previousSnapshot },
  } = useGame();
  const toast = useToast();

  const copyInviteLink = useCallback(async () => {
    try {
      await navigator.clipboard.writeText(`${window.location.href}/join`);

      toast({
        description: "Successfully copied invite link to clipboard",
        status: "success",
      });
    } catch (e) {
      toast({
        description:
          "Unable to copy invite link, try copying the current page URL directly",
        status: "error",
      });
    }
  }, [toast]);

  const canStartGame = useMemo(() => {
    const activePlayers = players.filter(
      ({ status }) => status !== PlayerStatus.SPECTATING,
    );

    return (
      activePlayers.length > 0 &&
      activePlayers.every(({ status }) => status === PlayerStatus.READY)
    );
  }, [players]);

  return (
    <Stack marginTop="50px" alignItems="center" spacing={8}>
      <HStack spacing={3} align="center">
        <Heading as="h1" textAlign="center">
          {gameName}
        </Heading>

        <Button
          leftIcon={<FaRegCopy />}
          width="fit-content"
          size="sm"
          onClick={() => void copyInviteLink()}
        >
          Copy invite link
        </Button>
      </HStack>

      <Stack
        alignItems={{ base: "center", md: "start" }}
        spacing={8}
        direction={{ base: "column", md: "row" }}
      >
        <Stack>
          <Button
            colorScheme="blue"
            isDisabled={!canStartGame}
            onClick={() => socket.emit(ClientToServerEventName.Split, null)}
          >
            Start game
          </Button>
          <PlayerTable />
        </Stack>

        {previousSnapshot && (
          <Stack>
            <Text textAlign="center">Here are the boards from that round:</Text>

            <OpponentBoardPreview
              initialPlayerIndex={previousSnapshot.findIndex(
                (player) => player.isTopBanana,
              )}
              players={previousSnapshot}
              includeCurrentPlayer
            />
          </Stack>
        )}
      </Stack>
    </Stack>
  );
}
