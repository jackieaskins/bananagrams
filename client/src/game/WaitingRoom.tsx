import { Col, Row, Space, Typography } from 'antd';

import PlayerTable from '../player/PlayerTable';
import { useGameName } from './stateHooks';

const { Text } = Typography;
const WaitingRoom = (): JSX.Element => {
  const gameName = useGameName();
  const joinUrl = `${window.location.href}/join`;

  return (
    <Row justify="center" css={{ marginTop: '25px' }}>
      <Col xs={20} sm={14} md={12} lg={10}>
        <Space direction="vertical" css={{ width: '100%' }} size="large">
          <div css={{ textAlign: 'center' }}>
            <h1>{gameName}</h1>
            <Text type="secondary">Invite others to game: </Text>
            <br />
            <Text code copyable type="secondary">
              {joinUrl}
            </Text>
          </div>

          <PlayerTable />
        </Space>
      </Col>
    </Row>
  );
};

export default WaitingRoom;
