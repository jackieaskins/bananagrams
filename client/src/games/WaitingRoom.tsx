import {
  Button,
  HStack,
  Heading,
  Stack,
  Text,
  useClipboard,
  useToast,
} from "@chakra-ui/react";
import { useEffect } from "react";
import { FaRegCopy } from "react-icons/fa6";
import OpponentBoardPreview from "../boards/OpponentBoardPreview";
import PlayerTable from "../players/PlayerTable";
import { useGame } from "./GameContext";

export default function WaitingRoom(): JSX.Element {
  const {
    gameInfo: { gameName, previousSnapshot },
  } = useGame();
  const joinUrl = `${window.location.href}/join`;
  const { onCopy, hasCopied } = useClipboard(joinUrl);
  const toast = useToast();

  useEffect(() => {
    if (hasCopied) {
      toast({ description: "Successfully copied invite link to clipboard" });
    }
  }, [hasCopied, toast]);

  return (
    <Stack marginTop="50px" alignItems="center" spacing={8}>
      <HStack spacing={3} align="center">
        <Heading as="h1" textAlign="center">
          {gameName}
        </Heading>

        <Button
          leftIcon={<FaRegCopy />}
          width="fit-content"
          size="sm"
          onClick={onCopy}
        >
          Copy invite link
        </Button>
      </HStack>

      <Stack
        alignItems={{ base: "center", md: "start" }}
        spacing={8}
        direction={{ base: "column", md: "row" }}
      >
        <PlayerTable />

        {previousSnapshot && (
          <Stack>
            <Text textAlign="center">Here are the boards from that round:</Text>

            <OpponentBoardPreview
              initialPlayerIndex={previousSnapshot.findIndex(
                (player) => player.isTopBanana,
              )}
              players={previousSnapshot}
              includeCurrentPlayer
            />
          </Stack>
        )}
      </Stack>
    </Stack>
  );
}
