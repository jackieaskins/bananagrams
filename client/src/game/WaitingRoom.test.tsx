import { shallow } from 'enzyme';

import WaitingRoom from './WaitingRoom';

jest.mock('./stateHooks', () => ({
  useGameName: jest.fn().mockReturnValue('gameName'),
}));

describe('<WaitingRoom />', () => {
  test('renders properly', () => {
    expect(shallow(<WaitingRoom />)).toMatchSnapshot();
  });
});
