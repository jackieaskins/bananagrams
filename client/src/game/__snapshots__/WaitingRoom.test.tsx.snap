// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WaitingRoom /> renders properly 1`] = `
.emotion-01r87dl5-WaitingRoom {
  margin-top: 25px;
}

.emotion-016tm0n-WaitingRoom {
  width: 100%;
}

.emotion-0iqsi0g-WaitingRoom {
  text-align: center;
}

<undefined
  className="emotion-0"
  justify="center"
>
  <Col
    lg={10}
    md={12}
    sm={14}
    xs={20}
  >
    <Space
      className="emotion-0"
      direction="vertical"
      size="large"
    >
      <div
        className="emotion-0"
      >
        <h1>
          gameName
        </h1>
        <Text
          type="secondary"
        >
          Invite others to game: 
        </Text>
        <br />
        <Text
          code={true}
          copyable={true}
          type="secondary"
        >
          http://localhost//join
        </Text>
      </div>
      <PlayerTable />
    </Space>
  </Col>
</undefined>
`;
